_posts/2019-05-17-State TTL in Flink 1.8.0- How to Automatically Cleanup Application State in Apache Flink.md

```diff
@@ -0,0 +1,136 @@
+
+```java
+StateTtlConfig ttlConfig = StateTtlConfig
+    .newBuilder(Time.days(7))
+    .cleanupFullSnapshot()
+    .build();
+```
+
+The local storage stays untouched but the size of the stored snapshot is reduced. The local state of an operator will only be cleaned up when the operator reloads its state from a snapshot, i.e. in case of recovery or when starting from a savepoint. 
+
+Due to these limitations, applications still need to actively remove state after it expired in Flink 1.6.0. To improve the user experience, Flink 1.8.0 introduces two more autonomous cleanup strategies, one for each of Flink’s two state backend types. We describe them below.
+
+## Incremental cleanup in Heap state backends
+
+This approach is specific to the Heap state backends (FSStateBackend and MemoryStateBackend). The idea is that the storage backend keeps a lazy global iterator over all state entries. Certain events, for instance state access, trigger an incremental cleanup. Every time an incremental cleanup is triggered, the iterator is advanced. The traversed state entries are checked and expired once are removed. The following code example shows how to enable incremental cleanup:
+
+```java
+StateTtlConfig ttlConfig = StateTtlConfig
+    .newBuilder(Time.days(7))
+    // check 10 keys for every state access
+    .cleanupIncrementally(10, false)
+    .build();
+```
+
+If enabled, every state access triggers a cleanup step. For every clean up step, a certain number of state entries are checked for expiration. There are two tuning parameters. The first defines the number of state entries to check for each cleanup step. The second parameter is a flag to trigger a cleanup step after each processed record, additionally to each state access.
+
+There are two important caveats about this approach: 
+* The first one is that the time spent for the incremental cleanup increases the record processing latency.
+* The second one should be practically negligible but still worth mentioning: if no state is accessed or no records are processed, expired state won’t be removed.
+
+## RocksDB background compaction to filter out expired state
+
+If your application uses the RocksDB state backend, you can enable another cleanup strategy which is based on a Flink specific compaction filter. RocksDB periodically runs asynchronous compactions to merge state updates and reduce storage. The Flink compaction filter checks the expiration timestamp of state entries with TTL and discards all expired values.
+
+The first step to activate this feature is to configure the RocksDB state backend by setting the following Flink configuration option:  
+`<state.backend.rocksdb.ttl.compaction.filter.enabled>`. Once the RocksDB state backend is configured, the compaction cleanup strategy is enabled for a state as shown in the following code example:
```

Comment:
I guess no newline before <state.backend.rocksdb.ttl.compaction.filter.enabled>, we can also drop `<>` because it is already highlighted with quotes `.